File: experimental/sso-with-skills/RootBot/TokenExchangeSkillHandler.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System;
using System.Globalization;
using System.Linq;
using System.Security.Claims;
using System.Security.Principal;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Integration.AspNet.Core.Skills;
using Microsoft.Bot.Builder.Skills;
using Microsoft.Bot.Connector.Authentication;
using Microsoft.Bot.Schema;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.Logging;
using Newtonsoft.Json.Linq;

namespace Microsoft.BotBuilderSamples.RootBot
{
    public class TokenExchangeSkillHandler : SkillHandler
    {
        private readonly BotAdapter _adapter;
        private readonly SkillsConfiguration _skillsConfig;
        private readonly SkillHttpClient _skillClient;
        private readonly string _botId;
        private readonly string _connectionName;
        private readonly SkillConversationIdFactoryBase _conversationIdFactory;
        private IExtendedUserTokenProvider _tokenExchangeProvider;

        public TokenExchangeSkillHandler(
            BotAdapter adapter,
            IBot bot,
            IConfiguration configuration,
            SkillConversationIdFactoryBase conversationIdFactory,
            SkillsConfiguration skillsConfig,
            SkillHttpClient skillClient,
            ICredentialProvider credentialProvider,
            AuthenticationConfiguration authConfig,
            IChannelProvider channelProvider = null,
            ILogger logger = null)
            : base(adapter, bot, conversationIdFactory, credentialProvider, authConfig, channelProvider, logger)
        {
            _adapter = adapter;
            _tokenExchangeProvider = adapter as IExtendedUserTokenProvider;
            if (_tokenExchangeProvider == null)
            {
                throw new ArgumentException($"{nameof(adapter)} does not support token exchange");
            }

            _skillsConfig = skillsConfig;
            _skillClient = skillClient;
            _conversationIdFactory = conversationIdFactory;

            _botId = configuration.GetSection(MicrosoftAppCredentials.MicrosoftAppIdKey)?.Value;
            _connectionName = configuration.GetSection("ConnectionName")?.Value;
        }

        protected override async Task<ResourceResponse> OnSendToConversationAsync(ClaimsIdentity claimsIdentity, string conversationId, Activity activity, CancellationToken cancellationToken = default(CancellationToken))
        {
            if (await InterceptOAuthCards(claimsIdentity, activity).ConfigureAwait(false))
            {
                return new ResourceResponse(Guid.NewGuid().ToString("N", CultureInfo.InvariantCulture));
            }

            return await base.OnSendToConversationAsync(claimsIdentity, conversationId, activity, cancellationToken).ConfigureAwait(false);
        }

        protected override async Task<ResourceResponse> OnReplyToActivityAsync(ClaimsIdentity claimsIdentity, string conversationId, string activityId, Activity activity, CancellationToken cancellationToken = default(CancellationToken))
        {
            if (await InterceptOAuthCards(claimsIdentity, activity).ConfigureAwait(false))
            {
                return new ResourceResponse(Guid.NewGuid().ToString("N", CultureInfo.InvariantCulture));
            }

            return await base.OnReplyToActivityAsync(claimsIdentity, conversationId, activityId, activity, cancellationToken).ConfigureAwait(false);
        }

        private BotFrameworkSkill GetCallingSkill(ClaimsIdentity claimsIdentity)
        {
            var appId = JwtTokenValidation.GetAppIdFromClaims(claimsIdentity.Claims);

            if (string.IsNullOrWhiteSpace(appId))
            {
                return null;
            }

            return _skillsConfig.Skills.Values.FirstOrDefault(s => string.Equals(s.AppId, appId, StringComparison.InvariantCultureIgnoreCase));
        }

        private async Task<bool> InterceptOAuthCards(ClaimsIdentity claimsIdentity, Activity activity)
        {
            var oauthCardAttachment = activity.Attachments?.FirstOrDefault(a => a?.ContentType == OAuthCard.ContentType);
            if (oauthCardAttachment != null)
            {
                var targetSkill = GetCallingSkill(claimsIdentity);
                if (targetSkill != null)
                {
                    var oauthCard = ((JObject)oauthCardAttachment.Content).ToObject<OAuthCard>();

                    if (!string.IsNullOrWhiteSpace(oauthCard?.TokenExchangeResource?.Uri))
                    {
                        using (var context = new TurnContext(_adapter, activity))
                        {
                            context.TurnState.Add<IIdentity>("BotIdentity", claimsIdentity);

                            // AAD token exchange
                            try
                            {
                                var result = await _tokenExchangeProvider.ExchangeTokenAsync(
                                    context,
                                    _connectionName,
                                    activity.Recipient.Id,
                                    new TokenExchangeRequest() { Uri = oauthCard.TokenExchangeResource.Uri }).ConfigureAwait(false);

                                if (!string.IsNullOrEmpty(result?.Token))
                                {
                                    // If token above is null, then SSO has failed and hence we return false.
                                    // If not, send an invoke to the skill with the token. 
                                    return await SendTokenExchangeInvokeToSkill(activity, oauthCard.TokenExchangeResource.Id, result.Token, oauthCard.ConnectionName, targetSkill, default).ConfigureAwait(false);
                                }
                            }
                            catch
                            {
                                // Show oauth card if token exchange fails.
                                return false;
                            }

                            return false;
                        }
                    }
                }
            }
            return false;
        }

        private async Task<bool> SendTokenExchangeInvokeToSkill(Activity incomingActivity, string id, string token, string connectionName, BotFrameworkSkill targetSkill, CancellationToken cancellationToken)
        {
            var activity = incomingActivity.CreateReply() as Activity;
            activity.Type = ActivityTypes.Invoke;
            activity.Name = SignInConstants.TokenExchangeOperationName;
            activity.Value = new TokenExchangeInvokeRequest()
            {
                Id = id,
                Token = token,
                ConnectionName = connectionName,
            };

            var skillConversationReference = await _conversationIdFactory.GetSkillConversationReferenceAsync(incomingActivity.Conversation.Id, cancellationToken).ConfigureAwait(false);
            activity.Conversation = skillConversationReference.ConversationReference.Conversation;
            activity.ServiceUrl = skillConversationReference.ConversationReference.ServiceUrl;

            // route the activity to the skill
            var response = await _skillClient.PostActivityAsync(_botId, targetSkill, _skillsConfig.SkillHostEndpoint, activity, cancellationToken);

            // Check response status: true if success, false if failure
            return (response.Status >= 200 && response.Status <= 299);
        }
    }
}

File: experimental/sso-with-skills/javascript_nodejs/rootBot/tokenExchangeSkillHandler.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

const { ActivityEx, CardFactory, SkillHandler, TurnContext } = require('botbuilder');
const { JwtTokenValidation } = require('botframework-connector');
const { ActivityTypes } = require('botframework-schema');
const { v4 } = require('uuid');

class TokenExchangeSkillHandler extends SkillHandler {
    constructor(
        adapter,
        bot,
        conversationIdFactory,
        skillsConfig,
        skillClient,
        credentialProvider,
        authConfig,
        channelService = null
    ) {
        super(adapter, bot, conversationIdFactory, credentialProvider, authConfig, channelService);

        if (!adapter) throw new Error('adapter is required for TokenExchangeSkillHandler');
        if (!skillsConfig) throw new Error('skillsConfig is required for TokenExchangeSkillHandler');
        if (!skillClient) throw new Error('skillClient is required for TokenExchangeSkillHandler');
        if (!conversationIdFactory) throw new Error('conversationIdFactory is required for TokenExchangeSkillHandler');

        this.adapter = adapter;
        this.skillsConfig = skillsConfig;
        this.skillClient = skillClient;
        this.conversationIdFactory = conversationIdFactory;

        this.botId = process.env.MicrosoftAppId;
        this.connectionName = process.env.ConnectionName;
    }

    async onSendToConversation(claimsIdentity, conversationId, activity) {
        if (await this.interceptOAuthCards(claimsIdentity, activity)) {
            return {
                id: v4()
            };
        }

        return await super.onSendToConversation(claimsIdentity, conversationId, activity);
    }

    async onReplyToActivity(claimsIdentity, conversationId, activityId, activity) {
        if (await this.interceptOAuthCards(claimsIdentity, activity)) {
            return {
                id: v4()
            };
        }

        return await super.onReplyToActivity(claimsIdentity, conversationId, activityId, activity);
    }

    getCallingSkill(claimsIdentity) {
        const appId = JwtTokenValidation.getAppIdFromClaims(claimsIdentity.claims);

        if (!appId) {
            return null;
        }

        return Object.values(this.skillsConfig.skills).find((s) => s.appId === appId);
    }

    async interceptOAuthCards(claimsIdentity, activity) {
        const oathCardAttachment = activity.attachments?.find((a) => a.contentType === CardFactory.contentTypes.oauthCard);
        if (oathCardAttachment) {
            const targetSkill = this.getCallingSkill(claimsIdentity);
            if (targetSkill) {
                const oauthCard = oathCardAttachment.content;

                if (oauthCard?.tokenExchangeResource?.uri) {
                    const context = new TurnContext(this.adapter, activity);

                    context.turnState.set('BotIdentity', claimsIdentity);

                    // AAD Token Exchange
                    try {
                        const result = await this.adapter.exchangeToken(
                            context,
                            this.connectionName,
                            activity.recipient.id,
                            { uri: oauthCard.tokenExchangeResource.uri }
                        );

                        if (result?.token) {
                            // If token above is null, then SSO has failed and hence we return false.
                            // If not, send an invoke to the skill with the token.
                            return this.sendTokenExchangeInvokeToSkill(activity, oauthCard.tokenExchangeResource.uri, result.token, oauthCard.connectionName, targetSkill);
                        }
                    } catch (err) {
                        // This will show oauth card if token exchange fails.
                        // A common cause for hitting this section of the code is when then user hasn't provided consent to the skill app.
                        console.warn(`Unable to get SSO token for OAuthCard, exception was ${ err }`);
                        return false;
                    }
                }
            }
        }

        return false;
    }

    async sendTokenExchangeInvokeToSkill(incomingActivity, id, token, connectionName, targetSkill) {
        const activity = ActivityEx.createReply(incomingActivity);
        activity.type = ActivityTypes.Invoke;
        activity.name = 'signin/tokenExchange';
        activity.value = {
            id,
            token,
            connectionName
        };

        const skillConversationReference = await this.conversationIdFactory.getSkillConversationReference(incomingActivity.conversation.id);
        activity.conversation = skillConversationReference.conversationReference.conversation;
        activity.serviceUrl = skillConversationReference.conversationReference.serviceUrl;

        // Route the activity to the skill.
        const response = await this.skillClient.postToSkill(this.botId, targetSkill, this.skillsConfig.skillHostEndpoint, activity);

        // Check response status: true if success, false if failure.
        return (response.status >= 200 && response.status <= 299);
    }
}

module.exports.TokenExchangeSkillHandler = TokenExchangeSkillHandler;
